presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java

```diff
@@ -0,0 +1,246 @@
+                        ((FunctionCall) (functionCall.getArguments().get(0))).getName().toString().equals("filter");
+            }
+
+            return false;
+        }
+
+        private static boolean isSimpleComaparison(ComparisonExpression comparisonExpression)
+        {
+            switch (comparisonExpression.getOperator()) {
+                case EQUAL:
+                case GREATER_THAN:
+                case LESS_THAN:
+                case NOT_EQUAL:
+                case GREATER_THAN_OR_EQUAL:
+                case LESS_THAN_OR_EQUAL:
+                    return true;
+            }
+
+            return false;
+        }
+
+        private Expression simplifyCardinalityOfFilterComparedToZero(
+                NodeLocation location,
+                Expression arrayArg,
+                ComparisonExpression.Operator operator,
+                LambdaArgumentDeclaration origLambdaArgument,
+                Expression origLambdaBody, String newLambdaArgumentName,
+                String finalLambdaArgumentName,
+                LambdaArgumentDeclaration finalLambdaArgument)
+        {
+            // Comparing to zero can be rewrtitten to a boolean.
+            // cardinality(filter(a, x -> f(x)) = 0 becomes reduce(a, false, (s, x) -> case when s then s else f(x) end, s -> !s)
+            // cardinality(filter(a, x -> f(x)) > 0 becomes reduce(a, false, (s, x) -> case when s then s else f(x) end, s -> s)
+            Expression finalLambdaBody = new Identifier(location, finalLambdaArgumentName, false);
+            ImmutableList.Builder<WhenClause> builder = new ImmutableList.Builder<WhenClause>();
+            builder.add(new WhenClause(new Identifier(location, newLambdaArgumentName, false), new Identifier(location, newLambdaArgumentName, false)));
+
+            if (operator == ComparisonExpression.Operator.EQUAL) {
+                // negate the result
+                finalLambdaBody = new NotExpression(finalLambdaBody);
+            }
+
+            LambdaArgumentDeclaration newLambdaArgument = new LambdaArgumentDeclaration(new Identifier(location, newLambdaArgumentName, false));
+            LambdaExpression newLambda = new LambdaExpression(location,
+                    ImmutableList.of(newLambdaArgument, origLambdaArgument),
+                    new SearchedCaseExpression(location, builder.build(), Optional.of(origLambdaBody)));
+            LambdaExpression finalLambda = new LambdaExpression(location, ImmutableList.of(finalLambdaArgument), finalLambdaBody);
+
+            // Now make the reduce
+            return new FunctionCall(
+                    location,
+                    QualifiedName.of("reduce"),
+                    ImmutableList.of(
+                            arrayArg,
+                            new BooleanLiteral("false"),
+                            newLambda,
+                            finalLambda));
+        }
+
+        private Expression simplifyCardinalityOfFilter(
```

Comment:
Let's also revamp this method to be in the following layout to facilitate review process, also, let's also rename the variables following the comments in `simplifyCardinalityOfFilterComparedToZero`:
```
- create cast (not required if it could be inlined)
- create case statement
- create input function
- create output function
- create reduce function
```